tools/ui/src/external.ts

```diff
@@ -0,0 +1,66 @@
+interface ExternalConfig {
+  candid?: string;
+}
+
+type MessageListener = (message: any) => void;
+
+const hasExternalConfig = new URLSearchParams(window.location.search).has(
+  "external-config"
+);
+
+const messagePrefix = "CandidUI";
+
+// Starting with a hard-coded origin list for security
+const allowedExternalOrigins = [
+  "http://ryjl3-tyaaa-aaaaa-aaaba-cai.localhost:8000", // Motoko Playground (local)
```

Comment:
This seems reasonable. `postMessage()` is fairly brittle in terms of security, so [most security audits require strict equality when matching domains](https://gist.github.com/jedp/3005816) because of all the ways people can imitate localhost URLs (e.g. registering a custom domain name that somehow matches the localhost regex). 

As a perfect example, when looking at the [MP external editor origin list](https://github.com/dfinity/motoko-playground/blob/4af2fcb110dd48a850658709a3b0ff635c883f71/src/integrations/allowedOrigins.js), I realized that there is actually a minor cross-origin vulnerability where someone could register a domain with anything in places of the "." symbols in `/^https?:\/\/(localhost|127.0.0.1)(:[0-9]+)?$/` (fixed by dfinity/motoko-playground#126). 

These cross-origin vulnerabilities tend to be extremely subtle, so we could probably air on the side of caution for the Candid UI for now (and maybe even scale back on the flexibility for the allowed MP origins).

Reply:
Sounds good.